clarifaipyspark/dataset.py

```diff
@@ -449,3 +453,77 @@ def export_dataset_to_dataframe(self, input_type, input_ids: list = None):
         how='left').drop(annotations_df.input_id)
+  
+  def export_annotations_to_volume(self, volumepath: str):
+    """Export all the annotations from clarifai App's dataset to UC volume along with annotated images.
+
+    Args:
+        volumepath (str): path of the volume storage where images will be downloaded.
+        
+    Examples:
+        TODO
+    
+    Returns:
+        None. Images are saved into the volume storage.
+    """
+
+    annotation_list = []
+    spark = SparkSession.builder.appName('Clarifai-spark').getOrCreate()
+
+    my_inputs = list(self.list_inputs(input_type='image'))
+    inp_dict=[]
+    for inp in my_inputs:
+      temp={}
+      temp['input_id']=inp.id
+      temp['image_url'] = inp.data.image.url
+      temp['img_format']= inp.data.image.image_info.format
+      inp_dict.append(temp)
+
+    response = list(self.list_annotations())
+    images_to_download=[]
+    for an in response:
+      temp = {}
+      temp['annotation'] = str(an.data)
+      if not temp['annotation'] or temp['annotation'] == '{}':
+        continue
+      temp['annotation_id'] = an.id
+      temp['annotation_user_id'] = an.user_id
+      temp['input_id'] = an.input_id
+      for val in inp_dict:
+          if an.input_id == val['input_id']:
+             temp['image_url'] = val['image_url']
+             if val not in images_to_download :
+                images_to_download.append(val)
+             
+      try:
+        created_at = float(f"{an.created_at.seconds}.{an.created_at.nanos}")
+        temp['annotation_created_at'] = time.strftime('%m/%d/% %H:%M:%5', time.gmtime(created_at))
+        modified_at = float(f"{an.modified_at.seconds}.{an.modified_at.nanos}")
+        temp['annotation_modified_at'] = time.strftime('%m/%d/% %H:%M:%5',
+                                                       time.gmtime(modified_at))
+      except:
+        temp['annotation_created_at'] = float(f"{an.created_at.seconds}.{an.created_at.nanos}")
+        temp['annotation_modified_at'] = float(f"{an.modified_at.seconds}.{an.modified_at.nanos}")
+      annotation_list.append(temp)
+    
+    df_delta = spark.createDataFrame(annotation_list)
+    df_delta.write.format("delta").mode("overwrite").save(volumepath)
+    df_url= pd.DataFrame(images_to_download)
+  
+    def download_image(args):
+      i, imgid, ext, url, volumepath= args
+      img_name = os.path.join(volumepath, f"{imgid}.{ext.lower()}")
+      headers = {"Authorization": self.metadata[0][1]}
+      response = requests.get(url, headers=headers)
+    
+      with open(img_name, "wb") as f:
+        f.write(response.content)
+    
+      return i
+    
+    args_list = [
+    (i, df_url.input_id[i], df_url.img_format[i], df_url.image_url[i], volumepath)
+    for i in range(df_url.shape[0])]
+
+    with ThreadPoolExecutor(max_workers=5) as executor:  # Adjust max_workers 
+      results = list(tqdm(executor.map(download_image, args_list), total=len(args_list), desc="Exporting Images"))
```

Comment:
Usage of tqdm is incorrect, progress should be updated when job is executed not when it is just submitted.
Could you update by using `as_completed` method

Reply:
and the comment `# Adjust max_workers` is not intuitive, how can users adjust this?

Reply:
Modified to use as_completed.